### PR TITLE
Fixing Issue #4 first try

### PR DESCRIPTION
**Pull request by Richardn2002**
磨刀不误砍柴功

不过。。
第一遍编译通过我就满意了
要是还能玩我就惊了

### CODE REVIEW

**Comment by Gennadiyev** [comment]
Start to compile the given codes!

**Comment by Gennadiyev** [comment]
Compilation Success! Testing the application

**Comment by Gennadiyev** [comment]
Issue found: **20 times click on "Change Mode" can be get repeatedly.

**Comment by Gennadiyev** [comment]
Issue found: **Looks like a variable is used twice. Will look into further.**

File: code/main.lua [hunk 1]
```diff
@@ -116,42 +220,94 @@ test.onClick=function(test)
...
+
+		--coming soon!
+
+		--end
+
+
+		--Thens
+		--4
+		if cancelling < 5 then
+			print("哇！居然点了取消！真没想到啊！那我就陪你玩玩好了！把程序名字设定为鲲的英语名试试？")
```
Comment (Gennadiyev) [inline code/main.lua]: `print("哇！居然点了取消！真没想到啊！那我就陪你玩玩好了！把程序名字设定为鲲的英语名试试"..string.rep("？",cancelling)) `

Comment (Gennadiyev) [reply code/main.lua]: This is to resolve Issue #7.

File: code/main.lua [hunk 2]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+	if mode==3 then
+		result.text=tostring(tonumber(adder1.text)/tonumber(adder2.text))
+	end
+
+
+	--Ifs
+
+	--0
+	if achievementCounter == 0 and mode == 3 and adder2.text == "0" and error_catch.checked then
+		achievementCounter = achievementCounter + 1
```
Comment (Gennadiyev) [inline code/main.lua]: `achievementCounter = 1` is a better choice. Optimized and clean. (Or you are thinking something evil...?)

File: code/main.lua [hunk 3]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+	end
+	--1
+	if notChanging == 0 and mode == 0 and adder1.text == "20" and adder2.text == "10" then
```
Comment (Gennadiyev) [inline code/main.lua]: Issue #12 Fixed here.

File: code/main.lua [hunk 4]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+		notChanging = notChanging + 1
```
Comment (Gennadiyev) [inline code/main.lua]: `notChanging = 1`  please.

File: code/main.lua [hunk 5]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+	end
+	--2
+	if hittingModes > 0 then
```
Comment (Gennadiyev) [inline code/main.lua]: In this way every single click on "CALC" would reset the counter. Infinity +1! That's a fatal error!
Use the following line to replace:

`if hittingModes > 0 and hittingModes < 20 then`

File: code/main.lua [hunk 6]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+	end
+	if fortyTwo == 1 and mode == 3 and tonumber(adder1.text)/tonumber(adder2.text) == 42 then
+		fortyTwo = fortyTwo + 1
+	end
+	--4
+	if cancelling > 0 then
+		cancelling = 0
+	end
+	--6
+	if calcAndCalc < 50 then
```
Comment (Gennadiyev) [inline code/main.lua]: It just feels strange... I prefer to erase this `if` and make `calcAndCalc = calcAndCalc + 1` always executed. Only the `calcAndCalc == 50` will be granted the achievement right? I bet nobody could get it to inf...

File: code/main.lua [hunk 7]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+		notChanging = notChanging + 1
+	end
+	--3
+	if fortyTwo == 2 then
+		found = found + 1
+		print("隐藏成就达成："..tostring(found).."/7")
+		adder1.text="SIX"
+		运算符.text="MULTIPLIED BY"
+		adder2.text="NINE"
+		fortyTwo = fortyTwo + 1
```
Comment (Gennadiyev) [inline code/main.lua]: Damn it. I want it to be `fortyTwo = 3`, please fix it.

File: code/main.lua [hunk 8]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+	end
+	--6
+	if calcAndCalc == 50 then
 		print("这是您第50次计算。那么，象征性给您成就半个！")
 		found=found+0.5
 		print("隐藏成就达成："..tostring(found).."/7")
+		calcAndCalc = calcAndCalc + 1
```
Comment (Gennadiyev) [inline code/main.lua]: `calcAndCalc = 51`. PLEASE!

File: code/main.lua [hunk 9]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+		knowingWhatKunWants = knowingWhatKunWants + 1
+	end
+
+
+	--Thens
+
+	--0
+	if achievementCounter == 1 then
+		print("ERROR: DIVIDED BY 0")
+		print("隐藏成就计数器："..tostring(found).."/7。")
```
Comment (Gennadiyev) [inline code/main.lua]: Your idea of the `ACHIEVEMENT_SUM` is great. Why not imply it?

Comment (Gennadiyev) [reply code/main.lua]: `print("隐藏成就计数器："..tostring(found).."/"..tostring(ACHIEVEMENT_SUM).."。") `

Comment (Gennadiyev) [reply code/main.lua]: Copy that for all the below please.

File: code/main.lua [hunk 10]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+		achievementCounter = 0
+	end
+	--1
+	if notChanging == 1 then
+		print("您就不能换一个数据嘛！我生气了！我要把答案搞成错误的！")
+		result.text="340"
```
Comment (Gennadiyev) [inline code/main.lua]: `result.text=tostring(WHAT_KUN_WANTS) `

File: code/main.lua [hunk 11]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+		print("隐藏成就达成："..tostring(found).."/7")
+		notChanging = notChanging + 1
+	end
+	--3
+	if fortyTwo == 2 then
+		found = found + 1
+		print("隐藏成就达成："..tostring(found).."/7")
+		adder1.text="SIX"
+		运算符.text="MULTIPLIED BY"
+		adder2.text="NINE"
```
Comment (Gennadiyev) [inline code/main.lua]: While most people can't simply search `6*9` online and get relevant result in short time, I would prefer `The ultimate answer` still be implied. Try to find another way? Developers don't want their game being stuck by players haven't been reading a book the dev has read.

File: code/main.lua [hunk 12]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
 	end
-	if ultimate==1 and adder2.text=="42" then
-		found = found + 1
+	--7
+	if knowingWhatKunWants > 0 and knowingWhatKunWants < 11 then
+		print("说，您调戏这个计算器多久了？那么熟练？")
+		knowingWhatKunWants = knowingWhatKunWants + 1
+	end
+	if knowingWhatKunWants > 10 then
+		print("不给不给不给"..string.rep("！",knowingWhatKunWants))
```
Comment (Gennadiyev) [inline code/main.lua]: But wait. The messages got way too long so easily, so I would prefer do this:
`print(string.rep("不给！",knowingWhatKunWants - 10)) `

File: code/main.lua [hunk 13]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+
+
+	--Ifs
+
+	--2
+	if hittingModes < 20 then
+		hittingModes = hittingModes + 1
+	end
+	--4
+	if cancelling > 0 then
```
Comment (Gennadiyev) [inline code/main.lua]: The `cancelling` achievement can be repeatedly acquired now. Fix this (similarly as the above).

File: code/main.lua [hunk 14]
```diff
@@ -8,91 +8,195 @@ activity.setTitle('鲲')
...
+		fortyTwo = fortyTwo + 1
+	end
+	if fortyTwo == 1 and mode == 2 and tonumber(adder1.text)*tonumber(adder2.text) == 42 then
+		fortyTwo = fortyTwo + 1
+	end
+	if fortyTwo == 1 and mode == 3 and tonumber(adder1.text)/tonumber(adder2.text) == 42 then
+		fortyTwo = fortyTwo + 1
+	end
+	--4
+	if cancelling > 0 then
```
Comment (Gennadiyev) [inline code/main.lua]: See line 169.

File: code/main.lua (hunk 13 above)
Comment (Gennadiyev) [reply code/main.lua]: *Actually I do not expect the five 'Cancel's to be consecutive...*

File: code/main.lua [hunk 15]
```diff
@@ -116,42 +220,94 @@ test.onClick=function(test)
...
-			print("哇！我们回来了！给你半个成就吧！")
-			found = found + 0.5
-			kun_is_back = 1
+
+
+		--Ifs
+
+		--5
+		if appname.text=="鲲" and kunIsBack == 0 then
+			kunIsBack = kunIsBack + 1
```
Comment (Gennadiyev) [inline code/main.lua]: I got too tired to complain... Okay it's your choice to ignore such lines.

File: code/main.lua (hunk 13 above)
Comment (Gennadiyev) [reply code/main.lua]: Oh yes, the corrected code:
`if cancelling > 0 and cancelling < 5 then`

**Review by Gennadiyev — CHANGES_REQUESTED** [review]
Nice work. Almost impossible for me without a compiler beside me! Thanks a lot w~

File: code/main.lua (hunk 2 above)
Comment (Richardn2002) [reply code/main.lua]: achievementCounter在条件满足时step+1，step==1触发，出发后清零，这样这个事件可以重复触发，没有什么问题
这样做是为了保持所有事件的结构一致

你写=1是因为你知道这个事件只需要1次条件符合就能触发，这很“特殊”，不“一般”

File: code/main.lua (hunk 4 above)
Comment (richard20020815) [reply code/main.lua]: see above

File: code/main.lua (hunk 11 above)
Comment (Richardn2002) [reply code/main.lua]: 我原来的意思是达成成就后变成6*9。。。可能我写错了

Comment (Gennadiyev) [reply code/main.lua]: Seems.

**Comment by richard20020815** [comment]
@Gennadiyev Done

File: code/main.lua (hunk 1 above)
Comment (Richardn2002) [reply code/main.lua]: Done

File: code/main.lua (hunk 11 above)
Comment (Richardn2002) [reply code/main.lua]: There is no error.

File: code/main.lua (hunk 2 above)
Comment (Richardn2002) [reply code/main.lua]: Now using -1 to mark completion.Closing all related complaints below.

File: code/main.lua (hunk 6 above)
Comment (Richardn2002) [reply code/main.lua]: This is for formatting reasons.

File: code/main.lua (hunk 3 above)
Comment (Richardn2002) [reply code/main.lua]: I didn't mean to...so we can learn that formatting does good!

File: code/main.lua (hunk 13 above)
Comment (Richardn2002) [reply code/main.lua]: Done.

Comment (Richardn2002) [reply code/main.lua]: All right.

Comment (Richardn2002) [reply code/main.lua]: Done.

File: code/main.lua (hunk 10 above)
Comment (Richardn2002) [reply code/main.lua]: Done.

File: code/main.lua (hunk 9 above)
Comment (Richardn2002) [reply code/main.lua]: Done.

File: code/main.lua (hunk 12 above)
Comment (Richardn2002) [reply code/main.lua]: Done.

**Comment by Richardn2002** [comment]
My FastHub App has gone mad.Those "Done" comments are messed up.

**Comment by Richardn2002** [comment]
oh no...there are some non-fatal bugs

I ctrl-vied too happily

**Comment by Richardn2002** [comment]
Comon code...plz work

**Review by Gennadiyev — APPROVED** [review]

File: code/main.lua [hunk 16]
```diff
@@ -116,42 +238,94 @@ test.onClick=function(test)
...
+		if appname.text=="Kunologist" and found ~= ACHIEVEMENT_SUM - 1 then
+			theEnd = -1
+		end
+
+		--Thens
+
+		--5
+		if kunIsBack == 1 then
+			found = found + 1
+			print("鲲又回来了")
```
Comment (Gennadiyev) [inline code/main.lua]: Wait since when did this achievement rewards 1 pt.?

**Pull request merged**